### ISSUE TITLE
[docs] CUDA generator output depends on the device's SM count: seeding does not make device-side sampling reproducible across GPU models or MIG slices

### ISSUE THREAD

**Issue by iamgahr** [pytorch/pytorch#196613]
### 📚 The doc issue

`torch.rand`/`torch.randn` (and every distribution kernel built on `distribution_nullary_kernel`) with a **seeded CUDA generator** produce different values on GPUs with a different number of SMs (or a different `maxThreadsPerMultiProcessor`) as soon as the tensor exceeds one wave of the launch grid, i.e. when

```
numel > 256 * multiProcessorCount * (maxThreadsPerMultiProcessor / 256)
```

This is by design and has been true for a long time — the RNG (Philox) is deterministic; what depends on the device is the mapping of output elements to Philox counters, because the grid is capped by the SM count:

```cpp
// aten/src/ATen/native/cuda/DistributionTemplates.h (v2.13.0), calc_execution_policy, lines 49-61
dim3 grid((numel + block_size - 1) / block_size);
uint32_t blocks_per_sm = at::cuda::getCurrentDeviceProperties()->maxThreadsPerMultiProcessor / block_size;
grid.x = std::min(static_cast<uint32_t>(at::cuda::getCurrentDeviceProperties()->multiProcessorCount) * blocks_per_sm, grid.x);
// lines 65-88: thread idx seeds curand_init(seed, idx, offset); element
// li = linear_index + blockDim.x*gridDim.x*ii takes component ii of that thread's k-th curand_uniform4()
```

But it is not stated where users look — the [Reproducibility note](https://pytorch.org/docs/stable/notes/randomness.html), `torch.Generator`, `torch.manual_seed`, `torch.cuda.manual_seed` — and the note's advice to seed the generator reads as if seeded device-side sampling were portable across GPUs. It is not: the same seed gives a different initialisation, a different dropout mask, a different augmentation sample on another SKU, on a MIG slice of the same card, or after a hardware refresh.

### What I measured (torch 2.13.0+cu129, `torch.Generator(device="cuda").manual_seed(0)`, `torch.rand(n)`, sha256 of the raw fp32 bytes)

| card (SMs × blocks/SM) | grid cap | `rand(2^16)` | `rand(2^20)` |
|---|---|---|---|
| L40 (142 × 6) | 852 | `2ccd6bcb5339…` | `28a2fbc8ed96…` |
| L40S (142 × 6) | 852 | `2ccd6bcb5339…` | `28a2fbc8ed96…` |
| H100 PCIe (114 × 8) | 912 | `2ccd6bcb5339…` | `ebd21badc8cf…` |
| A10G (80 × 6) | 480 | `2ccd6bcb5339…` | `ec3fe0ff87f0…` |

Below one wave (2^16 = 256 blocks) every card agrees; at 2^18 and above the L40, the H100 and the A10G all disagree, and so do the 1g.10gb / 3g.40gb MIG slices of one H100 against the whole card. A CPU re-implementation of Philox4x32-10 + cuRAND's `curand_init(seed, subsequence=idx, offset)` layout + `_curand_uniform` (integer → float first, then `x * 2^-32 + 2^-33`) + the grid formula above reproduces each of the three cards' output **bit for bit at every size tested (2^12 … 2^22)** when given that card's `(multiProcessorCount, maxThreadsPerMultiProcessor)`, and produces the other card's bytes when given the other card's numbers — so the dependence is exactly the launch geometry and nothing else. (`randn` shows the same card-dependence; I did not reproduce its Box–Muller bits.)

### Suggested change

A sentence in the Reproducibility note (and in the `torch.Generator` / `torch.cuda.manual_seed` docstrings), roughly:

> Values drawn from a CUDA generator depend on the device's multiprocessor count (and its maximum threads per multiprocessor) once a tensor is larger than one wave of the launch grid. Seeding does **not** make device-side sampling reproducible across different GPU models or MIG configurations. For cross-device reproducibility, sample on the CPU and copy to the device, or record the device properties alongside the seed.

I am not proposing to change the kernel — that would change every existing user's streams — only to say what it does.

### Versions

torch 2.13.0+cu129 (also observed with the cu126 build), driver 580.126.09 / 570.x; NVIDIA L40, L40S, H100 PCIe (and MIG slices), A100 SXM4, A10, A10G, L4, T4 — seven distinct (architecture, SM count) classes gave seven distinct `randn(2^20)` digests and the same digest whenever the SM count was the same.

<details><summary>CPU reproduction (numpy), ~60 lines</summary>

```python
import hashlib, numpy as np
M0, M1, W0, W1 = np.uint32(0xD2511F53), np.uint32(0xCD9E8D57), 0x9E3779B9, 0xBB67AE85
def mulhilo(a, b):
    p = np.uint64(a) * b.astype(np.uint64); return (p >> np.uint64(32)).astype(np.uint32), (p & np.uint64(0xFFFFFFFF)).astype(np.uint32)
def philox4x32_10(c0, c1, c2, c3, k0, k1):
    for _ in range(10):
        hi0, lo0 = mulhilo(M0, c0); hi1, lo1 = mulhilo(M1, c2)
        c0, c1, c2, c3 = hi1 ^ c1 ^ np.uint32(k0), lo1, hi0 ^ c3 ^ np.uint32(k1), lo0
        k0, k1 = (k0 + W0) & 0xFFFFFFFF, (k1 + W1) & 0xFFFFFFFF
    return c0, c1, c2, c3
def torch_cuda_rand(n, sm_count, max_threads_per_sm, seed=0):
    grid = min(sm_count * (max_threads_per_sm // 256), (n + 255) // 256); stride = 256 * grid
    n_iter = (n - 1) // (stride * 4) + 1
    idx = np.arange(stride, dtype=np.uint64); out = np.zeros(n_iter * stride * 4, dtype=np.float32)
    for k in range(n_iter):                       # curand_init(seed, idx, 0): ctr = (k, 0, idx_lo, idx_hi), key = seed
        r = philox4x32_10(np.full(stride, k, np.uint32), np.zeros(stride, np.uint32),
                          (idx & np.uint64(0xFFFFFFFF)).astype(np.uint32), (idx >> np.uint64(32)).astype(np.uint32),
                          seed & 0xFFFFFFFF, seed >> 32)
        for ii in range(4):                       # _curand_uniform: uint -> float first, then x*2^-32 + 2^-33
            xf = r[ii].astype(np.float32)
            v = (xf.astype(np.float64) * 2.0**-32 + 2.0**-33).astype(np.float32)
            v = np.where(v == np.float32(1.0), np.float32(0.0), v)      # uniform_kernel's reverse_bound_value
            out[k * stride * 4 + ii * stride + np.arange(stride)] = v
    return out[:n]
# torch side, on the GPU:  g = torch.Generator(device="cuda").manual_seed(0); t = torch.rand(1 << 20, device="cuda", generator=g)
#                          hashlib.sha256(t.cpu().numpy().tobytes()).hexdigest()
print(hashlib.sha256(torch_cuda_rand(1 << 20, 142, 1536).tobytes()).hexdigest()[:12])   # L40:  28a2fbc8ed96
print(hashlib.sha256(torch_cuda_rand(1 << 20, 114, 2048).tobytes()).hexdigest()[:12])   # H100: ebd21badc8cf
print(hashlib.sha256(torch_cuda_rand(1 << 20, 80, 1536).tobytes()).hexdigest()[:12])    # A10G: ec3fe0ff87f0
```
</details>

cc @svekars @sekyondaMeta @AlannaBurke @pbelevich

**Comment by iamgahr** [comment]
Update with the measurements since filing, and the threads this consolidates.

**1. The prediction is exact on four cards.** The CPU implementation of the mapping described above (Philox4x32-10 with cuRAND's `curand_init(seed, subsequence = thread index, offset)` layout, the grid capped at `multiProcessorCount × (maxThreadsPerMultiProcessor / 256)` blocks, `_curand_uniform`'s integer→float conversion) reproduces `torch.rand(n, device="cuda", generator=Generator(device="cuda").manual_seed(0))` **bit for bit** at n = 2^12, 2^14, 2^16, 2^18, 2^20 and 2^22 on every card we have run it against, given nothing but that card's SM count and threads per SM:

| card | SMs × blocks/SM (grid cap) | CPU prediction equals the device's `rand` |
|---|---|---|
| A10G | 80 × 6 = 480 | 6 of 6 sizes |
| L40 | 142 × 6 = 852 | 6 of 6 |
| L40S | 142 × 6 = 852 | 6 of 6 (and equals the L40 at every size — same SM count) |
| H100 PCIe | 114 × 8 = 912 | 6 of 6 |

Below one wave of threads (n ≤ 256 × SMs × blocks/SM) every card agrees with every other; above it each SM count is its own stream. Across an eleven-instance training campaign, `randn(2^20)` from a seeded device generator gave **seven distinct outputs for seven distinct (architecture, SM count) classes**, identical outputs for the same SM count (L40 = L40S; three H100 PCIe cards from three providers), and the MIG slices of one H100 (14 and 46 SMs) differ from the whole card.

**2. This is the mechanism behind three older threads.** #84234 (2022, "Randomness should be consistent across devices") asked whether there is "purposeful determinism in the mapping of array index to sequence number/counter" — there is not: the counter is assigned per *thread*, and the thread grid is a function of the SM count once the tensor exceeds one wave. #87992 (2022, `torch.rand` inconsistent across GPUs "once dimensions become large", from the diffusers tests) is the same threshold: 8 × 77 × 1024 = 630,784 elements is above one wave on every card that thread mentions. #194676 (2026) measured the boundary empirically on a T4 and an L4 and found it at exactly 40,960 = 256 × 40 SMs × 4 blocks/SM — one wave of the T4 — which is what the formula above gives.

**3. Two consequences worth stating in the docs, beyond "not guaranteed".** (a) A seeded device stream is *reproducible in principle on any machine*: given the SM count and threads-per-SM of the card that produced it, a CPU can regenerate it exactly (the script in this issue does), so "run on a different GPU" is a *transcription* problem, not a randomness problem — useful for anyone replaying or auditing a run. (b) If an opt-in were ever wanted, an element-indexed counter mapping (as discussed in #194676) would make seeded device sampling identical across GPU models by construction; the CPU implementation here is a ready-made specification and test oracle for it, since it states exactly which Philox (counter, key) each element consumes today.

**4. Context.** The other source of cross-GPU divergence in training — cuBLASLt selecting GEMM algorithms per SM count and per product — is localised and pinned in #105726 (comment of 2026-09-11): with the algorithm forced through the public cuBLASLt API, fp32 training is byte-identical across Ampere, Ada and Hopper and across MIG slices at the stock speed. With that and the launch-grid mapping above, every cross-device difference in our seven-class training measurements has a named cause; the only one without a user-side remedy today is this one, which is why the documentation sentence matters.
